Fail the integration tests on warnings or errors.

mum4k opened the pull request:
Fails the integration tests if any unknown warnings or errors are found in the logs of the Nighthawk test server.
Includes an ignore list that (for now) accepts known warnings found in the logs, the ignore list can have different content per each test case.

Also:
- moving the `Attributes:` docstring section on `IntegrationTestBase` into its class docstring. `__init__` is just a method and should only have the `Args:` section.

Fixes #577

Signed-off-by: Jakub Sobon <mumak@google.com>

mum4k commented:
@wjuan-AFK please review and assign back to me once done.

oschaaf commented on test/integration/nighthawk_test_server.py:
```diff
@@ -40,6 +40,29 @@ def _substitute_yaml_values(runfiles_instance, obj, params):
+    "Deprecated field: type envoy.config.filter.network.http_connection_manager.v2.HttpFilter",
+
+    # TODO(mum4k): Identify these and file issues or add explanation as necessary.
+    "Unable to use runtime singleton for feature envoy.http.headermap.lazy_map_min_size",
```
not actionable, but I think I remember issues being filed for this, it might be a bug in Envoy.

mum4k replied:
Thank you, if you do know the issue number, please let me know. Also your help in identifying background for the remaining ignored warnings would be much appreciated.

mum4k replied:
Note - we will be addressing some of the "Using deprecated extension name ..." warnings in my next PR. I will leave this in here as is for now, but leave the TODO and come back after the next PR gets merged in to see if additional cleanup is needed. Tracking in #582.

oschaaf commented on test/integration/BUILD:
```diff
@@ -9,6 +9,14 @@ licenses(["notice"])  # Apache 2
 envoy_package()
 
+py_test(
+    name = "test_nighthawk_test_server",
```
I wonder if tests for our test-code should go into a distinct place for clarity (arguably it's not test-code as we offer it for production purposes too, but still)

mum4k replied:
Makes sense, how about we create a subdirectory like `integration/unit_tests` ? Open to suggestions.

mum4k replied:
Addressed in 6cd4657596aa0a0d383bfd63d65995a30c9aa84a.

oschaaf commented on benchmarks/envoy_proxy.py:
```diff
@@ -92,9 +92,9 @@ def setUp(self):
     self.proxy_server = proxy_server
 
-  def tearDown(self):
+  def tearDown(self, caplog):
```
This adds to the benchmarking framework, specifically when Envoy proxy gets injected between the client and test server. Is the plan to add to the other fixtures in a follow up? E.g. https://github.com/envoyproxy/nighthawk/blob/master/test/integration/integration_test_fixtures.py ?

mum4k replied:
I thought this is what this PR was doing. We are only adding that parameter to the benchmarking framework, because it inherits from the same fixtures. This PR does modify `integration_test_fixtures.py` and that is where the failure on errors / warnings happens. Is this what you meant or am I missing something?

oschaaf replied:
You are right, sorry about the noise, please disregard.

oschaaf reviewed (commented):
looks great, just two questions/thoughts.

mum4k commented:
Adding @dubious90 in case he can get to this sooner than @wjuan-AFK.

oschaaf reviewed (dismissed):
LGTM

wjuan-AFK commented:
LGTM

wjuan-AFK commented on test/integration/integration_test_fixtures.py:
```diff
@@ -134,6 +140,15 @@ def tearDown(self):
         any_failed = True
     assert (not any_failed)
 
+    messages = []
```
Maybe this name could be slightly clearer about what kind of messages they are? Probably not necessary.

mum4k replied:
Addressed in 61821dbaeca4c7cddf6923cc38f0f693b760ca37.

mum4k commented:
@oschaaf please take another look after addressing feedback and merging from master. Sorry about the force push, I have fat-fingered a bad rebase into this branch and had to rewrite history to fix it. Same commits were re-applies as cherry-picks.

mum4k commented:
@oschaaf please take another look after merging from master.

mum4k commented:
@oschaaf fyi this PR needs bit more work. Now that we do have a test which verifies we pass with `v2` config if the user requests it, we fail because that warning appears in the server logs. Going to update this so that we can filter out messages per specific set of test cases.

mum4k commented:
@oschaaf this is ready for another review. We can now specify distinct ignore lists based on the name of the executed test case. The test case that purposefully runs with the deprecated Envoy v2 API is now allowed to log the relevant warnings.

mum4k commented:
@wjuan-AFK significant changes were made since your last review to allow this to pass tests after #584 was merged in. For your convenience, [here is a view](https://github.com/envoyproxy/nighthawk/pull/581/files/2c3597657f93b3253074759391159502a6106ccd..e10c7b368f47bd174dec4f4b37610fe8f0b2a880) that only includes commits related to this change.

oschaaf reviewed (approved):
Nice, this will be very useful, LGTM

wjuan-AFK commented:
LGTM
